packages/loopring_v3/contracts/amm/libamm/AmmExchange.sol

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/ERC20.sol";
+import "../../lib/MathUint.sol";
+import "./AmmData.sol";
+import "./AmmPoolToken.sol";
+import "./AmmUtil.sol";
+
+
+/// @title AmmExchange
+library AmmExchange
+{
+    using AmmPoolToken      for AmmData.State;
+    using MathUint          for uint;
+
+    // Only used to withdraw from the pool when shutdown.
+    // Otherwise LPs should withdraw by doing normal queued exit requests.
+    function withdrawFromPoolWhenShutdown(
+        AmmData.State storage S,
+        uint                  poolAmountIn
+        )
+        public
+    {
+        require(poolAmountIn > 0, "INVALID_POOL_AMOUNT");
+        require(S.balanceOf[msg.sender] >= poolAmountIn, "INSUFFCIENT_POOL_AMOUNT");
+
+        uint size = S.tokens.length;
+        uint32 accountID = S.accountID;
+        IExchangeV3 exchange = S.exchange;
+        bool withdrawalMode = exchange.isInWithdrawalMode();
+
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+
+            // Question(Brecht): I removed a "!" isWithdrawnInWithdrawalMode, is it now correct?
+            require(
+                withdrawalMode && exchange.isWithdrawnInWithdrawalMode(accountID, token) ||
+                !withdrawalMode && !exchange.isForcedWithdrawalPending(accountID, token),
+                "PENDING_WITHDRAWAL"
+            );
```

Comment:
Looks correct, find this a bit more complicated to understand without the if on withdrawalMode.